src/source/Ice/IceAgent.c

```diff
@@ -695,6 +715,145 @@ STATUS iceAgentShutdown(PIceAgent pIceAgent)
+    pIceAgent->iceAgentStatus = STATUS_SUCCESS;
+    pIceAgent->lastDataReceivedTime = INVALID_TIMESTAMP_VALUE;
+
+    pIceAgent->relayCandidateCount = 0;
+
+    CHK_STATUS(doubleListGetHeadNode(pIceAgent->localCandidates, &pCurNode));
+    while (pCurNode != NULL) {
+        pLocalCandidate = (PIceCandidate) pCurNode->data;
+        pCurNode = pCurNode->pNext;
+
+        if (pLocalCandidate->iceCandidateType == ICE_CANDIDATE_TYPE_RELAYED) {
+            CHK_STATUS(turnConnectionShutdown(pLocalCandidate->pTurnConnection, 0));
+        }
+        localCandidates[localCandidateCount++] = pLocalCandidate;
+    }
+    CHK_STATUS(doubleListClear(pIceAgent->localCandidates, FALSE));
+
+    /* free all candidate pairs except the selected pair */
+    CHK_STATUS(doubleListGetHeadNode(pIceAgent->iceCandidatePairs, &pCurNode));
+    while (pCurNode != NULL) {
+        pIceCandidatePair = (PIceCandidatePair) pCurNode->data;
+        pNextNode = pCurNode->pNext;
+
+        if (pIceCandidatePair != pIceAgent->pDataSendingIceCandidatePair) {
+            CHK_STATUS(freeIceCandidatePair(&pIceCandidatePair));
+        }
+
+        pCurNode = pNextNode;
+    }
+    CHK_STATUS(doubleListClear(pIceAgent->iceCandidatePairs, FALSE));
+
+    MUTEX_UNLOCK(pIceAgent->lock);
+    locked = FALSE;
+
+    /* Time given for turn to free its allocation */
+    THREAD_SLEEP(HUNDREDS_OF_NANOS_IN_A_SECOND);
+
+    /* At this point there should be no thread accessing anything in iceAgent other than
+     * pIceAgent->pDataSendingIceCandidatePair and its ice candidates. Therefore safe to proceed freeing resources */
+
+    for (i = 0; i < localCandidateCount; ++i) {
+        if (localCandidates[i] != pIceAgent->pDataSendingIceCandidatePair->local) {
+            if (localCandidates[i]->iceCandidateType != ICE_CANDIDATE_TYPE_RELAYED) {
+                CHK_STATUS(connectionListenerRemoveConnection(pIceAgent->pConnectionListener, localCandidates[i]->pSocketConnection));
+                CHK_STATUS(freeSocketConnection(&localCandidates[i]->pSocketConnection));
+            } else {
+                CHK_STATUS(freeTurnConnection(&localCandidates[i]->pTurnConnection));
+            }
+            MEMFREE(localCandidates[i]);
+        }
+    }
+
+    /* Do not free remoteCandidates because new remote candidates could be added before or while restart happens.
+     * There is no way to tell which session a remote candidate belongs to. Old ones will eventually fail the
+     * connectivity test so it's ok. */
+
+    CHK_STATUS(stackQueueClear(pIceAgent->triggeredCheckQueue, FALSE));
+
+    ATOMIC_STORE_BOOL(&pIceAgent->remoteCredentialReceived, FALSE);
+    ATOMIC_STORE_BOOL(&pIceAgent->agentStartGathering, FALSE);
+    ATOMIC_STORE_BOOL(&pIceAgent->candidateGatheringFinished, FALSE);
+
+    pIceAgent->stateEndTime = 0;
+    pIceAgent->foundationCounter = 0;
+    pIceAgent->localNetworkInterfaceCount = ARRAY_SIZE(pIceAgent->localNetworkInterfaces);
```

Comment:
Not sure if this is worth it, but maybe we can regather the network interfaces with `getLocalhostIpAddresses` in case that there's a new network interface?

Reply:
yh it does do regathering. localNetworkInterfaceCount is set to ARRAY_SIZE(pIceAgent->localNetworkInterfaces); because getLocalhostIpAddresses() expects the second param to contain the array size.

Reply:
Ah, makes sense. I forgot that at the end of the call, it'll call `iceAgentStartGathering`.